### PR TITLE
[dalgona] Introduce Dalgona class

### PR DESCRIPTION
**Pull request by jinevening**
This introduces Dalgona class, which invokes luci-interpreter.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/ONE/pull/3501

### CODE REVIEW

**Comment by jinevening** [comment]
For reviewers:

Most of the codes are the same with `record-minmax`. A major difference will be Observer of Interpreter.

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/dalgona/src/Dalgona.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+using Shape = std::vector<loco::Dimension>;
+using DataType = loco::DataType;
+
+namespace
+{
+
+uint32_t numElements(const luci::CircleNode *node)
+{
+  uint32_t num_elements = 1;
```
Comment (mhs4670go) [inline compiler/dalgona/src/Dalgona.cpp:34]: How about add `assert` or throw an error for the case the `node` is `nullptr`?

File: compiler/dalgona/src/Dalgona.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  for (uint32_t i = 0; i < node->rank(); i++)
+    num_elements *= node->dim(i).value();
+
+  return num_elements;
+}
+
+// Return tensor's size in bytes
+template <typename NodeT> size_t getByteSize(const NodeT *node)
```
Comment (mhs4670go) [inline compiler/dalgona/src/Dalgona.cpp:42]: ditto.

File: compiler/dalgona/src/Dalgona.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+{
+  uint32_t dtype_size = loco::size(node->dtype());
+  return dtype_size * numElements(node);
+}
+
+// Throw exception if input has one of the following conditions.
+// 1. Have unknown dimension
+// 2. Number of elements is 0
+void checkInputDimension(const luci::CircleInput *input)
```
Comment (mhs4670go) [inline compiler/dalgona/src/Dalgona.cpp:51]: ditto

File: compiler/dalgona/src/Dalgona.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+          importer.readTensor(record_idx, input_idx, &dtype, &shape, input_data.data());
+
+          // Check the type and the shape of the input data is valid
+          verifyTypeShape(input_node, dtype, shape);
+        }
+        else
+        {
+          // Skip type/shape check for raw data
+          importer.readTensor(record_idx, input_idx, input_data.data());
+        }
```
Comment (mhs4670go) [inline compiler/dalgona/src/Dalgona.cpp:159]: How about use positive condition for better readability?
```suggestion
        if (is_raw_data)
        {
          // Skip type/shape check for raw data
          importer.readTensor(record_idx, input_idx, input_data.data());
        }
        else
        {
          DataType dtype;
          Shape shape;
          importer.readTensor(record_idx, input_idx, &dtype, &shape, input_data.data());

          // Check the type and the shape of the input data is valid
          verifyTypeShape(input_node, dtype, shape);
        }
```

**Comment by mhs4670go** [comment]
(suggestion for future) Current design forces to write each methods according to the format of input data file - HDF5(`runAnalysisWithH5Input`), RANDOM(`runAnalysisWithRandomInput`). If we split apart the section of loading data, it would have better extensibility like `circle-eval-diff` does.

**Comment by mhs4670go** [comment]
(optional) It would be better to add unittest for methods in unnamed namespace later - `numElements`, `getByteSize`, `checkInputDimension`, `verifyTypeShape`.

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM

**Pull request merged**